master/buildbot/test/integration/test_latent.py

```diff
@@ -0,0 +1,224 @@
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+from buildbot.config import BuilderConfig
+from buildbot.process.factory import BuildFactory
+from buildbot.test.fake.reactor import NonThreadPool
+from buildbot.test.fake.reactor import TestReactor
+from buildbot.test.util.integration import getBuilderIdByName
+from buildbot.test.util.integration import getMaster
+from buildbot.worker.base import AbstractLatentWorker
+
+from twisted.internet.defer import Deferred
+from twisted.python import threadpool
+from twisted.python.failure import Failure
+from twisted.trial.unittest import SynchronousTestCase
+
+
+class LatentController(object):
+    """
+    A controller for ``ControllableLatentWorker``.
+    """
+
+    def __init__(self, name):
+        self.worker = ControllableLatentWorker(name, self)
+        self.started = False
+
+    def start_instance(self, result):
+        assert self.started
+        self.started = False
+        d, self._start_deferred = self._start_deferred, None
+        d.callback(result)
+
+
+class ControllableLatentWorker(AbstractLatentWorker):
+    """
+    A latent worker that can be contolled by tests.
+    """
+
+    def __init__(self, name, controller):
+        AbstractLatentWorker.__init__(self, name, None)
+        self._controller = controller
+
+    def checkConfig(self, name, _):
+        AbstractLatentWorker.checkConfig(self, name, None)
+
+    def reconfigService(self, name, _):
+        AbstractLatentWorker.reconfigService(self, name, None)
+
+    def start_instance(self, build):
+        self._controller.started = True
+        self._controller._start_deferred = Deferred()
+        return self._controller._start_deferred
+
+    def stop_instance(self, build):
+        return Deferred()
+
+
+class TestException(Exception):
+    """
+    An exception thrown in tests.
+    """
+
+
+class Tests(SynchronousTestCase):
+
+    def setUp(self):
+        self.patch(threadpool, 'ThreadPool', NonThreadPool)
+        self.reactor = TestReactor()
+
+    def test_refused_substantiations_get_requeued(self):
+        """
+        If a latent worker refuses to substantiate, the build request becomes unclaimed.
+        """
+        controller = LatentController('local')
+        config_dict = {
+            'builders': [
+                BuilderConfig(name="testy",
+                              workernames=["local"],
+                              factory=BuildFactory(),
+                              ),
+            ],
+            'workers': [controller.worker],
+            'protocols': {'null': {}},
+            'multiMaster': True,
```

Comment:
I assume you use this to disable additional coherency tests (no scheduler for testy).
 I think this would make sense to have a 'disableCoherencyChecks': True, and make 'multiMaster' an alias for that.\